backend/src/main/java/com/example/demo/service/ProductServiceImpl.java

```
@@ -0,0 +1,214 @@
+import java.util.List;
+import java.util.Optional;
+
+import org.springframework.stereotype.Service;
+
+import com.example.demo.dto.Metrics;
+import com.example.demo.model.Product;
+import com.example.demo.repository.ProductRepository;
+
+@Service
+public class ProductServiceImpl implements ProductService {
+
+    private final ProductRepository pr;
+
+    public ProductServiceImpl(ProductRepository pr) {
+        this.pr = pr;
+    }
+
+    //IMPLEMENTED METHODS
+    //CREATE PRODUCT
+    @Override
+    public Product createProduct(Product product) {
+        boolean savedValue = pr.saveNewProduct(product);
+
+        if (!savedValue) {
+            throw new IllegalArgumentException("This product already exists");
+        }
+
+        product.setCreationDate(LocalDate.now());
+        product.setUpdateDate(LocalDate.now());
+
+        return product;
+    }
+
+    //UPDATE PRODUCT
+    @Override
+    public Product updateProduct(Integer id, Product product) {
+        boolean savedValue = pr.updateById(id, product);
+
+        if (!savedValue) {
+            throw new IllegalArgumentException("This product doesn't exist");
+        }
+
+        pr.updateById(id, product);
+        product.setUpdateDate(LocalDate.now());
+        return product;
+    }
+
+    //DELETE PRODUCT
+    @Override
+    public Product deleteById(Integer id) {
+        Product product = pr.findProductById(id);
+
+        if (product == null) {
+            throw new IllegalArgumentException("This product doesn't exist");
+        } else {
+            pr.deleteById(id);
+            return product;
+        }
+    } 
+
+    //MARK OUT OF STOCK
+    @Override
+    public Product markOutOfStock(Integer id) {
+        Product product = pr.findProductById(id);
+        Integer stock = product.getQuantityInStock();
+        if (stock != 0) {
+            product.setQuantityInStock(0);
+        } else {
+            throw new IllegalStateException("This product is already out of stock");
```

do you need to throw an exception here?
is it part of the project specs?
if not you could avoid the else

---

done! decided to avoid it